documentation/intermediate-ca.md

```diff
@@ -0,0 +1,191 @@
+# Intermediate CA support with Puppet Server
+
+## Configuration guide
+
+When Puppet Server or PE is installed it will autogenerate a number of certificates. Regenerating all of these certificates can be a time consuming process, but this problem can be side stepped by generating an intermediate CA certificate that reuses the existing key pair that Puppet Server/PE has already created.
+
+### Configure the Puppet agent `certificate_revocation` checking
+
+Intermediate CA support is incompatible with full chain revocation checking, as Puppet is unable to load multiple CRLs. When an intermediate CA certificate is in use Puppet must have the `certificate_revocation` setting either set to `false`, or `leaf`. By default Puppet will attempt to perform full chain checking and will fail when encountering a non-root CA certificate, so this setting **must** be reconfigured on all Puppet agents before continuing.
+
+    puppet config set --section main certificate_revocation leaf
+
+See [PUP-3788](https://tickets.puppetlabs.com/browse/PUP-3788) for more information on chain revocation checking.
+
+### Generate an intermediate CA certificate request
+
+    #!/opt/puppetlabs/puppet/bin/ruby
+    require 'puppet'
+    require 'openssl'
+
+    class Puppet::Application::GenCACSR < Puppet::Application
+      run_mode :master
+
+      def main
+        root_cert = OpenSSL::X509::Certificate.new(File.read(Puppet[:cacert]))
+        ca_key = OpenSSL::PKey::RSA.new(File.read(Puppet[:cakey]))
+
+        csr = OpenSSL::X509::Request.new
+        csr.version = 0
+        csr.subject = root_cert.subject
+        csr.public_key = root_cert.public_key
+
+        ef = OpenSSL::X509::ExtensionFactory.new
+
+        ext_req = OpenSSL::X509::Attribute.new("extReq",
+          OpenSSL::ASN1::Set([OpenSSL::ASN1::Sequence([
+            ef.create_extension("keyUsage", "cRLSign, keyCertSign", true),
+            ef.create_extension("basicConstraints", "CA:TRUE", true)])]))
+
+        csr.add_attribute(ext_req)
+        csr.sign(ca_key, OpenSSL::Digest::SHA256.new)
+
+        puts csr.to_s
+      end
+    end
+    Puppet::Application::GenCACSR.new.run
+
+Use this CSR to generate the intermediate CA certificate that Puppet will be using.
+
+### Construct the CA certificate chain
+
+Puppet Server needs to present the full certificate chain to clients in order for the client to authenticate the server. The certificate chain is constructed by concatenating the CA certificates together, starting with the new intermediate CA certificate and descending to the root CA certificate.
+
+    cat puppet-intermediate-certificate.pem \
```

Comment:
I can get it from context, but we should probably include where each of these files came from. Ie,
```
In this example, puppet-intermediate-certificate.pem is the output of the previous ruby script. You can find it at <default location on disk>.
org-intermediate-certificate.pem and org-root-certificate.pem are certificates in use by your organization.
```

Reply:
I like this system where you write good documentation and I just copy/paste it into the docs. Let's keep doing this!

Otherwise phrased, that's an excellent point and I'll add that.